blockchain/contracts/Staking.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import "@openzeppelin/contracts/interfaces/IERC721.sol";
+import "@openzeppelin/contracts/interfaces/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract Staking is IERC721Receiver {
+    // TODO natspec
+    // TODO emergencyWithdrawal, allocateRewards
+    IERC721 public characterToken;
+    IERC20 public gemsToken;
+    uint256 public STAKING_FEE = 100 * 10**18;
+
+    struct Stake {
+        uint256 tokenId;
+        uint256 gemsAmount;
+        uint256 timestamp;
+    }
+
+    event NFTStaked(address owner, uint256 tokenId, uint256 gemsAmount);
+    event NFTUnstaked(address owner, uint256 tokenId, uint256 rewardedAmount);
+    event Claimed(address owner);
+
+    // Map staker address to stake details
+    mapping(address => Stake) public stakes;
+    // Easier to manage than verifying the content of the Stake Struct which can be tricky due to deletion
+    mapping(address => bool) public isStaking;
```

Comment:
Another way I've seen this done is having an `isInitialized` boolean on the struct that is set to true for valid structs and defaults to false if there is no reference for that address

Reply:
Just sharing - up to you what to do here

Reply:
Sounds definitely better, will update